File: 15.三数之和.go
/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 */
package leetcode
import "sort"

// @lc code=start
func threeSum(nums []int) [][]int {
	var res [][]int
	if len(nums) < 3 {
		return res
	}
	sort.Ints(nums)
	i := 0
	for i < len(nums) - 2 {
		j, k := i + 1, len(nums) - 1
		for j < k {
			if nums[i] + nums[j] + nums[k] == 0 {
				res = append(res, []int{nums[i], nums[j], nums[k]})
				for  j + 1 < k && nums[j] == nums[j + 1] {
					j++
				}
				j++
			} else if nums[i] + nums[j] + nums[k] < 0 {
				for j + 1 < k && nums[j + 1] == nums[j] {
					j++
				}
				j++
			} else {
				for k - 1 > j && nums[k] == nums[k-1] {
					k--
				}
				k--
			}
		}
		for i + 1 < len(nums) - 2 && nums[i] == nums[i + 1] {
			i++
		}
		i++
	}
	return res
}
// @lc code=end

File: 15.三数之和.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 */
#include <algorithm>
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        int i = 0, j = nums.size() - 1;
        if (nums[i] > 0 || nums[j] < 0 || nums.size() < 3)
            return {};
        for (i = 0; i < nums.size() - 2; i++)
        {
            if (nums[i] > 0)
                return res;
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            int j = i + 1, k = nums.size() - 1;
            while (j < k)
            {
                int temp = nums[i] + nums[j] + nums[k];
                if (temp > 0)
                {
                    k--;
                }
                else if (temp < 0)
                {
                    j++;
                } else {
                    res.push_back({nums[i], nums[j], nums[k]});
                    while (j + 1 < k && nums[j + 1] == nums[j])
                        j++;
                    while (k - 1 > j && nums[k - 1] == nums[k])
                        k--;
                    j++; 
                    k--;
                }
            }
        }
        return res;
    }
};
// @lc code=end
